### ISSUE TITLE
Add an option to build ASTs for files in the CDB on startup

### ISSUE THREAD

**Issue by daandemeyer** [clangd/clangd#589]
Currently, clangd builds the AST for a file and its headers when it first opened. This takes a noticeable amount of time in large codebases. It would be great to have an option to have clangd do this on startup along with background indexing for all files in the compilation database.

**Comment by DaanDeMeyer** [comment]
Prototype patch implementing a --prebuild-asts option: https://reviews.llvm.org/D91509

**Comment by sam-mccall** [comment]
Yeah, new file latency is the worst. And we don't have big plans to fundamentally reduce it.

Can you explain a bit more about the workflow where you think this will help? e.g. typical TU takes A seconds to build, you expect to mark B TUs in your project as "pre-build" of which you'll end up opening C on average, clangd will typically run for D hours. Do you close inactive files and typically keep E open at once, or does that number grow over time?

The proposed solution seems pretty narrow: it brings benefits when TUs are slow to build, but also has high costs in that case. We may take some latency off the user's critical path, but:
 - peg the CPU and IO on startup, every time. (unlike background indexing, high priority threads)
 - lock out real interactive work (worker threads share a semaphore)
 - not work well with multiple-editor workflows
 - not work well if the editor is frequently closed/reopered
 - waste RAM/disk for any unopened files
It requires manual configuration for the user's particular workflow (and maintenance of this over time), so it's unlikely to be used by many people. Accordingly, it's hard to justify the complexity of a more robust implementation (separate thread pools etc).

Had a brief chat with @kadircet @kirillbobyrev, and we had a couple of ideas:
 - have a **persistent** preamble cache on disk, rather than considering these files private to the session. This trades latency for disk pretty predictably.
 - have a more relaxed eviction policy for preambles, so we should only build each preamble once per session. (This doesn't fix the latency, but does more to ensure it's amortized over long sessions)
(Focus on preambles, where the vast majority of AST build time goes and which are serializable. Also on measures that can be controlled with a general policy, rather than configuration of specific files)

There's also a low-tech/short term solution of forcing the relevant files to actually be opened on startup with a script in the editor. Certainly not ideal, but cheap if we can't find a solution that's useful to many people...

**Comment by DaanDeMeyer** [comment]
>peg the CPU and IO on startup, every time. (unlike background indexing, high priority threads)

Surely there's a way to mark the threads doing AST prebuilding as low pri, no?

The rest of the disadvantages are valid, but if we mention these explicitly and make it opt-in, why not give users the choice? There's the additional maintenance cost of course but the implementation is pretty simple and shouldn't have a big maintenance overhead.

I'm all for better solutions like the ones you mentioned but I'm assuming this isn't on the short-term roadmap for the team and unfortunately, I don't have the cycles required to implement a better solution at the moment. I can always keep this around as a patch since it seems like no one's complained before so there might not be a big ask from users for a feature like this.

**Comment by ArcsinX** [comment]
> * have a **persistent** preamble cache on disk, rather than considering these files private to the session. This trades latency for disk pretty predictably.

Seems preamble on disk is not enough. We also need other fields of `PreambleData` structure (`Diags`, `Includes`, etc.) + index update (which happens during preamble build).

**Comment by DaanDeMeyer** [comment]
I got all the bugs ironed out and this is working as expected. clangd takes about 5 minutes to build index and prebuild ASTs. After that, even when accessing the huge translation units I have almost instant access to the AST.

**Comment by HighCommander4** [comment]
>     * have a more relaxed eviction policy for preambles, so we should only build each preamble once per session. (This doesn't fix the latency, but does more to ensure it's amortized over long sessions)

That would be nice. I often close editor tabs just to manage tab clutter, without considering that I'm throwing away ASTs that will need to be rebuilt.

**Comment by DaanDeMeyer** [comment]
I abandoned the diff after trying it out for a while as it takes clangd too long to start up even with a limited amount of translation units when using it.

**Comment by sam-mccall** [comment]
@ArcsinX 

> Seems preamble on disk is not enough. We also need other fields of PreambleData structure (Diags, Includes, etc.) + index update (which happens during preamble build).

This is a really good point, and makes this a much larger project.
 - diags could be omitted if we treat preambles as stale (they're likely to be) with no diags. Maybe we can live without statcache and a few others for similar reasons.
 - but many other fields need to be serialized
 - and index is the killer: code completion won't work without it. And IIRC deserializing a full preamble to index it is the same order of magnitude as building it.

One big solution to this is making the *dynamic index* persistent too (at least the preamble parts of it): the dynamic index and preamble metadata could both live in a sqlite database (and overlap more with the background index).

@DaanDeMeyer 

That's a shame, but not all that surprising as we're not fundamentally able to skip any of the slow work.

**Comment by bronikkk** [comment]
@DaanDeMeyer Can you please tell me which project with huge translation units did you test on? Currently, we have a similar problem of generating a  preamble cache in our project.

**Comment by bronikkk** [comment]
@DaanDeMeyer Sorry to bother you, but do you have the diff that you abandoned?
